### PR TITLE
rt refine+haskell: prove tc_corres_caps and tc_corres_sched

### PR DESCRIPTION
**Pull request by ryybrr**
Prove `tc_corres_caps` and `tc_corres_sched`. Some changes to the specs based on what I've proved as part of my patch. Priority proofs are still sorried. Some adjustments to the main two lemmas will be needed soon.

### CODE REVIEW

File: proof/refine/ARM/Tcb_R.thy [hunk 1]
```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
...
+  done
+
+lemma getScTime_wp:
+  "\<lbrace>\<lambda>s. \<forall>rt. (\<forall>tcb p sc. (ko_at' tcb t s \<and> tcbSchedContext tcb = Some p \<and> ko_at' sc p s
+                                                          \<longrightarrow> rTime (refillHd sc) = rt)) \<longrightarrow> Q rt s\<rbrace>
+   getScTime t \<lbrace>Q\<rbrace>"
+  by (wpsimp wp: threadGet_wp simp: getScTime_def getTCBSc_def obj_at'_def)
+
+schematic_goal mapM_getScTime_wp:
+  "\<lbrace>?P\<rbrace> mapM getScTime ts \<lbrace>Q\<rbrace>"
```
Comment (corlewis) [inline proof/refine/ARM/Tcb_R.thy:2306]: It's a bit unusual to have a specific `mapM` lemma for a constant, but no problem if it makes other proofs easier.

File: proof/refine/ARM/Tcb_R.thy [hunk 2]
```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
...
+  apply (clarsimp simp: image_def in_set_conv_decomp zip_append1)
+  apply (rule_tac x="hd (drop (length ys) rvs)" in exI)
+  apply (case_tac "drop (length ys) rvs"; fastforce dest: list_all2_lengthD)
+  done
+
+crunches postpone, schedContextResume
+  for invs': invs'
+  (wp: crunch_wps)
+
+lemma threadSet_mdb':
```
Comment (corlewis) [inline proof/refine/ARM/Tcb_R.thy:2377]: This might be useful in an earlier file with other `threadSet` lemmas.

File: proof/refine/ARM/Tcb_R.thy [hunk 3]
```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
...
+    apply (wpsimp simp: C)+
+         apply (rule hoare_FalseE)
+        apply (rule hoare_FalseE)
+       apply (rule hoare_FalseE)
+      apply (wpsimp wp: B isFinal simp: capRemovable_def finaliseCap_def isEndpointCap_def)+
+   apply (fastforce simp: C cte_wp_at'_def final_matters'_def
+                          capRemovable_def invs'_def valid_state'_def)+
+  done
+
+lemma hoare_vcg_common_conj_liftE:
```
Comment (corlewis) [inline proof/refine/ARM/Tcb_R.thy:2539]: FIXME: move

**Review by corlewis — APPROVED** [review]

**Review by mktnk3 — APPROVED** [review]
This is very cool. Thanks for this work, Ryan. I am not sure where you are at with the remaining part of the work, but if you think this can be merged first, please go ahead.

File: spec/haskell/src/SEL4/Kernel/Thread.lhs [hunk 4]
```diff
@@ -526,22 +526,7 @@ The following function is used to alter the priority of a thread.
 >     case ts of
 >          Running -> threadSetPriority_onRunning tptr prio
 >          Restart -> threadSetPriority_onRunning tptr prio
->          BlockedOnSend ep _ _ _ _ -> threadSetPriority_onEp tptr prio ep
->          BlockedOnReceive ep _ _ -> threadSetPriority_onEp tptr prio ep
```
[the remark this answers is not in the document]
Comment (ryybrr) [reply spec/haskell/src/SEL4/Kernel/Thread.lhs:530]: I will implement this in my priority PR

**Review by michaelmcinerney — APPROVED** [review]
Cool

**Review by MitchellBuckley — APPROVED** [review]

File: proof/refine/ARM/Tcb_R.thy [hunk 5]
```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
...
+                          is_valid_vtable_root_def valid_ipc_buffer_cap tcb_ep_slot_cte_wp_at
+                          cte_wp_at_disj cte_wp_at_eq_simp real_cte_at_cte  real_cte_at_not_tcb_at
+                   split: option.split
+          | intro conjI | fastforce simp: valid_fault_handler_def)+)[1]
+  apply (clarsimp simp: tcb_at_cte_at'_0 tcb_at_cte_at'_1 tcb_at_cte_at'_3
+                        tcb_at_cte_at'_4 isCap_simps case_option_If2
+                        isValidFaultHandler_def isValidVTableRoot_def | intro conjI)+
+  done
+
+lemma get_sc_corres':
```
[the remark this answers is not in the document]
Comment (ryybrr) [reply proof/refine/ARM/Tcb_R.thy:1828]: I believe `get_sc_corres_size` conveys more information, so I'll use that one instead.

File: proof/refine/ARM/KHeap_R.thy [hunk 6]
```diff
@@ -134,6 +134,31 @@ lemma no_fail_getObject_tcb [wp]:
...
+  apply (simp add: lookupAround2_known1)
+  done
+
+lemma objBitsKO_bounded[simp]:
+  "objBitsKO ko < word_bits"
+  using scBits_max
+  by (simp add: objBits_simps' word_bits_def pageBits_def archObjSize_def pdeBits_def pteBits_def
+           split: Structures_H.kernel_object.split arch_kernel_object.split)
+
+lemma no_fail_getObject_sc[wp]:
```
Comment (michaelmcinerney) [inline proof/refine/ARM/KHeap_R.thy:150]: Don't we already have a `no_fail` rule for this or something very similar? `no_fail_getSchedContext`?

File: proof/refine/ARM/Tcb_R.thy [hunk 7]
```diff
@@ -1963,768 +1943,1232 @@ where
...
+            getThreadVSpaceRoot_def getThreadCSpaceRoot_def installTCBCap_def
+  by (wpsimp wp: checkCap_inv crunch_wps assertDerived_wp_weak | intro conjI)+
+
+lemma cteInsert_sa_simple[wp]:
+  "cteInsert newCap srcSlot destSlot \<lbrace>sch_act_simple\<rbrace>"
+  by (simp add: sch_act_simple_def, wp)
+
+lemma installTCBCap_sch_act_simple:
+  "\<lbrace>invs' and sch_act_simple and tcb_at' a\<rbrace>
+    installTCBCap a sl n sl_opt
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Tcb_R.thy:2061]: Indentation

File: proof/refine/ARM/Tcb_R.thy [hunk 8]
```diff
@@ -1963,768 +1943,1232 @@ where
...
+                        getReleaseQueue_def setReleaseQueue_def setReprogramTimer_def)
+  apply (wpsimp wp: threadSet_tcbInReleaseQueue_invs')
+     apply (clarsimp simp: ksReleaseQueue_ksReprogramTimer_update
+                           ksPSpace_ksReprogramTimer_update if_cancel_eq_True)
+     apply (wpsimp wp: mapM_wp_lift getScTime_wp threadGet_wp)+
+  apply (clarsimp simp: invs'_def valid_state'_def comp_def obj_at'_def)
+  apply (rule conjI)
+   apply (clarsimp simp: valid_release_queue_def obj_at'_def projectKO_eq projectKO_tcb objBitsKO_def)
+   apply (intro conjI impI; clarsimp)
+       apply (auto split: if_splits elim: ps_clear_domE)[3]
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Tcb_R.thy:2769]: Please remove these autos if at all possible

File: proof/refine/ARM/Tcb_R.thy (hunk 3 above)
Comment (mktnk3) [reply proof/refine/ARM/Tcb_R.thy:2947]: Can you please move this rather than labeling it?

File: proof/refine/ARM/Tcb_R.thy [hunk 9]
```diff
@@ -1963,768 +1943,1232 @@ where
...
-          apply wp
-         apply (wpsimp wp: cteDelete_invs' hoare_vcg_conj_lift)
-        apply (fastforce simp: emptyable_def)
-       apply fastforce
+    apply (subgoal_tac "obj_at' (\<lambda>sc. \<exists>y. scTCB sc = Some y) scp s'")
+     apply (clarsimp simp: sc_at_pred_def obj_at_def obj_at'_def)
+    apply (prop_tac "sc_at scp s")
+     apply (clarsimp simp: sc_at_pred_def obj_at_def obj_at'_def is_sc_obj_def)
+     apply (fastforce dest: invs_valid_objs simp: valid_obj_def)
+    apply (drule sc_at_cross[rotated 3], fastforce, fastforce, fastforce)
```
Comment (mktnk3) [inline proof/refine/ARM/Tcb_R.thy:2320]: It's not really any different, but this can be `apply (drule sc_at_cross[rotated 3]; fastforce?)` I think.

**Review by mktnk3 — APPROVED** [review]
This looks great. Thanks a lot for finishing this through to the end.

There is one request: we had some discussion and it was confirmed that the corres rules should have the concrete side function in their names (well, I was the one who wasn't clear on it). Shall we say that we follow that rule for naming the corres rules that we add newly for MCS? We probably should wait to rename the existing ones, but we can be at least consistent in adding new ones.

File: proof/refine/ARM/Tcb_R.thy (hunk 9 above)
Comment (ryybrr) [reply proof/refine/ARM/Tcb_R.thy:2320]: Doesn't work here

Comment (ryybrr) [reply proof/refine/ARM/Tcb_R.thy:2320]: Condensed this line and the following lines into a single fastforce instance

**Comment by mktnk3** [comment]
Looks all good! Please go ahead and merge, along with the stateAssert one, when you are ready.

**Comment by mktnk3** [comment]
Hmm. It looks like you need to sort out the Refine failure. I guess you can merge the stateAssert one in the meantime.

**Comment by mktnk3** [comment]
@lsf37 the Refine test here is failing but we have no idea why. It finishes fine when tested locally. And the error message is odd, there isn’t a “done” in the line it’s referring to.

**Comment by lsf37** [comment]
> @lsf37 the Refine test here is failing but we have no idea why. It finishes fine when tested locally. And the error message is odd, there isn’t a “done” in the line it’s referring to.

Does look fishy. I have restarted the test in case it's a fluke.

**Comment by lsf37** [comment]
Since this is development branch, it'd also be sure to just merge if you're certain it works locally. If the test keeps failing afterwards, we can still sort it out then. In the worst case the test will then be failing for everyone for a while, though.

**Comment by mktnk3** [comment]
> > @lsf37 the Refine test here is failing but we have no idea why. It finishes fine when tested locally. And the error message is odd, there isn’t a “done” in the line it’s referring to.
> 
> Does look fishy. I have restarted the test in case it's a fluke.

I should have mentioned. This was the second run. We got this error twice.

**Comment by mktnk3** [comment]
> Since this is development branch, it'd also be sure to just merge if you're certain it works locally. If the test keeps failing afterwards, we can still sort it out then. In the worst case the test will then be failing for everyone for a while, though.

Yes, we are not dependent on the tests on the PR, so we are ok in terms of merging and getting on with things.

One question I can ask is how one would investigate such errors. Anything that I can do?

**Pull request merged**